File: MinJarLaunch/launcher.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
#include <windows.h>
#include <shellapi.h>
#else
#include <unistd.h>
#endif

#ifdef _WIN32
static void showErrorAndExit(const char* message) {
    int len = MultiByteToWideChar(CP_ACP, 0, message, -1, NULL, 0);
    wchar_t* wMessage = (wchar_t*)malloc(len * sizeof(wchar_t));
    MultiByteToWideChar(CP_ACP, 0, message, -1, wMessage, len);

    MessageBoxW(NULL, wMessage, L"Error!", MB_ICONERROR | MB_OK);
    free(wMessage);
    exit(EXIT_FAILURE);
}
#else
static void showErrorAndExit(const char* message) {
    fprintf(stderr, "%s\n", message);
    exit(EXIT_FAILURE);
}
#endif

static int is_java_in_path() {
#ifdef _WIN32
    char buffer[MAX_PATH];
    DWORD result = SearchPath(NULL, "java.exe", NULL, MAX_PATH, buffer, NULL);
    return result > 0;
#else
    const char* path = getenv("PATH");
    if (path == NULL) {
        return 0;
    }

    char* path_copy = strdup(path);
    char* token = strtok(path_copy, ":");
    while (token != NULL) {
        char java_path[4096];
        snprintf(java_path, sizeof(java_path), "%s/java", token);
        if (access(java_path, X_OK) == 0) {
            free(path_copy);
            return 1;
        }
        token = strtok(NULL, ":");
    }
    free(path_copy);
    return 0;
#endif
}

#ifdef _WIN32
int WINAPI wWinMain(
    _In_ HINSTANCE hInstance,
    _In_opt_ HINSTANCE hPrevInstance,
    _In_ PWSTR szCmdLine,
    _In_ int cmdShow) {

    if (is_java_in_path()) {
        int argc;
        LPWSTR* argv = CommandLineToArgvW(GetCommandLineW(), &argc);

        if (argc > 1) {
            char command[4096];
            int ret = snprintf(command, sizeof(command), "javaw -jar \"%S\"", argv[1]);
            if (ret < 0 || ret >= sizeof(command)) {
                showErrorAndExit("Command buffer overflow!");
            }

            for (int i = 2; i < argc; i++) {
                ret = snprintf(command + strlen(command), sizeof(command) - strlen(command), " \"%S\"", argv[i]);
                if (ret < 0 || ret >= sizeof(command) - strlen(command)) {
                    showErrorAndExit("Command buffer overflow!");
                }
            }

            STARTUPINFO si;
            PROCESS_INFORMATION pi;
            ZeroMemory(&si, sizeof(si));
            si.cb = sizeof(si);
            ZeroMemory(&pi, sizeof(pi));

            if (!CreateProcess(NULL, command, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi)) {
                showErrorAndExit("javaw.exe not found!");
            }

            WaitForSingleObject(pi.hProcess, INFINITE);
            CloseHandle(pi.hProcess);
            CloseHandle(pi.hThread);
        }
        else {
            MessageBoxW(NULL, L"No file specified!", L"Info!", MB_ICONEXCLAMATION | MB_OK);
            LocalFree(argv);
            return EXIT_FAILURE;
        }

        LocalFree(argv);
    }
    else {
        MessageBoxW(NULL, L"JAVA not found!\n\nDownload and install a JRE.", L"Info!", MB_ICONEXCLAMATION | MB_OK);
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
#else
int main(int argc, char* argv[]) {
    if (is_java_in_path()) {
        if (argc > 1) {
            char* javaCmd[argc + 2];
            javaCmd[0] = "java";
            javaCmd[1] = "-jar";
            for (int i = 1; i < argc; i++) {
                javaCmd[i + 1] = argv[i];
            }
            javaCmd[argc + 1] = NULL;

            execvp("java", javaCmd);
            perror("Error: java not found or error executing command");
            return EXIT_FAILURE;
        }
        else {
            fprintf(stderr, "No file specified!\n");
            return EXIT_FAILURE;
        }
    }
    else {
        fprintf(stderr, "JAVA not found!\n\nDownload and install a JRE.\n");
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
#endif

File: MinJarLaunch/launcher.rs
use std::env;
use std::ffi::OsString;
use std::process::{Command, exit};
use std::ptr::null_mut;

#[cfg(windows)]
use std::os::windows::ffi::OsStringExt;
#[cfg(windows)]
use winapi::um::shellapi::CommandLineToArgvW;
#[cfg(windows)]
use winapi::um::winuser::{MessageBoxW, MB_ICONERROR, MB_ICONEXCLAMATION, MB_OK};
#[cfg(windows)]
use winapi::um::processthreadsapi::{CreateProcessW, PROCESS_INFORMATION, STARTUPINFOW};
#[cfg(windows)]
use winapi::um::handleapi::CloseHandle;
#[cfg(windows)]
use winapi::um::synchapi::WaitForSingleObject;
#[cfg(windows)]
use winapi::um::winbase::INFINITE;

#[cfg(windows)]
fn show_error_and_exit(message: &str) {
    let wide_message: Vec<u16> = OsString::from(message).encode_wide().chain(Some(0)).collect();
    unsafe {
        MessageBoxW(null_mut(), wide_message.as_ptr(), wide_message.as_ptr(), MB_ICONERROR | MB_OK);
    }
    exit(1);
}

#[cfg(unix)]
fn show_error_and_exit(message: &str) {
    eprintln!("{}", message);
    exit(1);
}

fn is_java_in_path() -> bool {
    if cfg!(windows) {
        let output = Command::new("cmd").args(&["/C", "where java"]).output();
        output.map_or(false, |o| o.status.success())
    } else {
        let output = Command::new("sh").arg("-c").arg("which java").output();
        output.map_or(false, |o| o.status.success())
    }
}

#[cfg(windows)]
fn main() {
    if is_java_in_path() {
        let cmd_line = unsafe { CommandLineToArgvW(winapi::um::processenv::GetCommandLineW(), &mut 0) };
        if cmd_line.is_null() {
            show_error_and_exit("Failed to get command line arguments");
        }

        let args: Vec<OsString> = unsafe {
            let mut args = Vec::new();
            let mut i = 0;
            while !(*cmd_line.offset(i)).is_null() {
                args.push(OsString::from_wide(std::slice::from_raw_parts(*cmd_line.offset(i), 1)));
                i += 1;
            }
            args
        };

        if args.len() > 1 {
            let mut command = OsString::from("javaw -jar ");
            command.push(&args[1]);

            for arg in &args[2..] {
                command.push(" ");
                command.push(arg);
            }

            let mut si: STARTUPINFOW = unsafe { std::mem::zeroed() };
            si.cb = std::mem::size_of::<STARTUPINFOW>() as u32;
            let mut pi: PROCESS_INFORMATION = unsafe { std::mem::zeroed() };

            let success = unsafe {
                CreateProcessW(
                    null_mut(),
                    command.encode_wide().chain(Some(0)).collect::<Vec<u16>>().as_mut_ptr(),
                    null_mut(),
                    null_mut(),
                    false as i32,
                    0,
                    null_mut(),
                    null_mut(),
                    &mut si,
                    &mut pi,
                )
            };

            if success == 0 {
                show_error_and_exit("javaw.exe not found!");
            }

            unsafe {
                WaitForSingleObject(pi.hProcess, INFINITE);
                CloseHandle(pi.hProcess);
                CloseHandle(pi.hThread);
            }
        } else {
            show_error_and_exit("No file specified!");
        }
    } else {
        show_error_and_exit("JAVA not found!\n\nDownload and install a JRE.");
    }
}

#[cfg(unix)]
fn main() {
    if is_java_in_path() {
        let args: Vec<String> = env::args().collect();
        if args.len() > 1 {
            let mut java_cmd = vec!["java".to_string(), "-jar".to_string()];
            java_cmd.extend_from_slice(&args[1..]);

            let status = Command::new("java").args(&java_cmd).status();
            if let Err(e) = status {
                eprintln!("Error: java not found or error executing command: {}", e);
                exit(1);
            }
        } else {
            eprintln!("No file specified!");
            exit(1);
        }
    } else {
        eprintln!("JAVA not found!\n\nDownload and install a JRE.");
        exit(1);
    }
}
